src/test/osd/RadosModel.h

```diff
@@ -1948,6 +1958,224 @@ class CopyFromOp : public TestOp {
   }
 };
 
+class SetRedirectOp : public TestOp {
+public:
+  string oid, oid_tgt, tgt_pool_name;
+  ObjectDesc src_value, tgt_value;
+  librados::ObjectWriteOperation op;
+  librados::ObjectReadOperation rd_op;
+  librados::AioCompletion *comp;
+  ceph::shared_ptr<int> in_use;
+  int done;
+  int r;
+  SetRedirectOp(int n,
+	     RadosTestContext *context,
+	     const string &oid,
+	     const string &oid_tgt,
+	     const string &tgt_pool_name,
+	     TestOpStat *stat = 0)
+    : TestOp(n, context, stat),
+      oid(oid), oid_tgt(oid_tgt), tgt_pool_name(tgt_pool_name),
+      comp(NULL), done(0), 
+      r(0)
+  {}
+
+  void _begin() override
+  {
+    Mutex::Locker l(context->state_lock);
+    context->oid_in_use.insert(oid);
+    context->oid_not_in_use.erase(oid);
+    context->oid_redirect_in_use.insert(oid_tgt);
+    context->oid_redirect_not_in_use.erase(oid_tgt);
+
+    context->find_object(oid, &src_value); 
+    if(!context->redirect_objs[oid].empty()) {
+      /* update target's user_version */
+      rd_op.stat(NULL, NULL, NULL);
+      comp = context->rados.aio_create_completion();
+      context->io_ctx.aio_operate(context->prefix+oid_tgt, comp, &rd_op,
+			    librados::OPERATION_ORDER_READS_WRITES,
+			    NULL);
+      comp->wait_for_safe();
+      context->update_object_version(oid_tgt, comp->get_version64());
+      comp->release();
+
+      /* unset redirect target */
+      comp = context->rados.aio_create_completion();
+      bool present = !src_value.deleted();
+      context->remove_object(oid);
+      op.remove();
+      context->io_ctx.aio_operate(context->prefix+oid, comp, &op,
+				  librados::OPERATION_ORDER_READS_WRITES |
+				  librados::OPERATION_IGNORE_REDIRECT);
+      comp->wait_for_safe();
+      if ((r = comp->get_return_value())) {
+	if (!(r == -ENOENT && !present)) {
+	  cerr << "r is " << r << " while deleting " << oid << " and present is " << present << std::endl;
+	  ceph_abort();
+	}
+      }
+      comp->release();
+
+      context->oid_redirect_not_in_use.insert(context->redirect_objs[oid]);
+      context->oid_redirect_in_use.erase(context->redirect_objs[oid]);
+
+      /* copy_from oid_tgt --> oid */
+      comp = context->rados.aio_create_completion();
+      context->find_object(oid_tgt, &tgt_value);
+      string src = context->prefix+oid_tgt;
+      op.copy_from(src.c_str(), context->io_ctx, tgt_value.version);
+      context->io_ctx.aio_operate(context->prefix+oid, comp, &op,
+				  librados::OPERATION_ORDER_READS_WRITES);
+      comp->wait_for_safe();
+      if ((r = comp->get_return_value())) {
+	cerr << "Error: oid " << oid << " copy_from " << oid_tgt << " returned error code "
+	     << r << std::endl;
+	ceph_abort();
+      }
+      context->update_object_full(oid, tgt_value);
+      context->update_object_version(oid, comp->get_version64());
+      comp->release();
+    }
```

Comment:
Nope. UnsetRedirect just remove a object. But this block will not only remove a object but also create a new object (through copy_from).